tests/example.c

```diff
@@ -95,27 +95,27 @@ int main()
         printf(" 8. print users\n");
         printf(" 9. count users\n");
         printf("10. quit\n");
-        gets(in);
+        fgets(in, sizeof(in), stdin);
         switch(atoi(in)) {
             case 1:
                 printf("name?\n");
-                add_user(id++, gets(in));
+                add_user(id++, fgets(in, sizeof(in), stdin));
                 break;
             case 2:
                 printf("id?\n");
-                gets(in);
+                fgets(in, sizeof(in), stdin);
                 id = atoi(in);
                 printf("name?\n");
-                add_user(id, gets(in));
+                add_user(id, fgets(in, sizeof(in), stdin));
```

Comment:
I do like the idea of eliminating `gets` here — not so much for the buffer-overflow issue, as because it is no longer standard C. (It was removed in C11.) However:

Just as `fputs`' behavior is different from `puts`', `fgets`' behavior is different from `gets`'. Specifically, `gets` trims the trailing newline, but `fgets` does not. (Autobiographical note: [I once wrote some code](http://www.club.cc.cmu.edu/~ajo/free-software/getline/getline.html) to work around the quirks here. But I'm not pulling in that library just for this toy example!)

So this patch is changing behavior — if the user enters `Arthur`, it'll be stored in the hash as `"Arthur\n"`. Which is... fine, I guess... but it feels wrong. Could we perhaps factor out two helper functions `int get_int()` and `const char *get_name(char *buf, int n)`, and use them throughout? How much uglier would that make the code?

Reply:
> I do like the idea of eliminating `gets` here — not so much for the buffer-overflow issue, as because it is no longer standard C. (It was removed in C11.) However:
> 
> Just as `fputs`' behavior is different from `puts`', `fgets`' behavior is different from `gets`'. Specifically, `gets` trims the trailing newline, but `fgets` does not. (Autobiographical note: [I once wrote some code](http://www.club.cc.cmu.edu/~ajo/free-software/getline/getline.html) to work around the quirks here. But I'm not pulling in that library just for this toy example!)

Removing a newline can be done with less code then described in  [I once wrote some code](http://www.club.cc.cmu.edu/~ajo/free-software/getline/getline.html).
> 
> So this patch is changing behavior — if the user enters `Arthur`, it'll be stored in the hash as `"Arthur\n"`. Which is... fine, I guess... but it feels wrong. Could we perhaps factor out two helper functions `int get_int()` and `const char *get_name(char *buf, int n)`, and use them throughout? How much uglier would that make the code?

I agree, will fix that.